docs/enhancements/20190724-prometheus-refactoring.md

```diff
@@ -0,0 +1,370 @@
+
+**Type:** gauge
+
+**Labels:**
+
+- controller_pod
+
+----------
+
+**Name:** nginx_ingress_controller_ingress_upstream_latency_seconds
+
+**Description:** Upstream service latency per Ingress
+
+**Type:** summary
+
+**Labels:**
+
+- controller_pod
+- ingress_uid
+
+----------
+
+**Name:** nginx_ingress_controller_nginx_process_connections
+
+**Description:** Current number of client connections with state {reading, writing, waiting}
+
+**Type:** gauge
+
+**Labels:**
+
+- controller_pod
+- state (reading, waiting, writing)
+
+----------
+
+**Name:** nginx_ingress_controller_nginx_process_connections_total
+
+**Description:** Total number of connections with state {active, accepted, handled}
+
+**Type:** counter
+
+**Labels:**
+
+- controller_pod
+- state (accepted, active, handled)
+
+----------
+
+**Name:** nginx_ingress_controller_nginx_process_cpu_seconds_total
+
+**Description:** CPU usage in seconds
+
+**Type:** counter
+
+**Labels:**
+
+- controller_pod
+
+----------
+
+**Name:** nginx_ingress_controller_nginx_process_num_procs
+
+**Description:** Number of processes
+
+**Type:** gauge
+
+**Labels:**
+
+- controller_pod
+
+----------
+
+**Name:** nginx_ingress_controller_nginx_process_oldest_start_time_seconds
+
+**Description:** Start time in seconds since 1970/01/01
+
+**Type:** gauge
+
+**Labels:**
+
+- controller_pod
+
+----------
+
+**Name:** nginx_ingress_controller_nginx_process_read_bytes_total
+
+**Description:** Number of bytes read
+
+**Type:** counter
+
+**Labels:**
+
+- controller_pod
+
+----------
+
+**Name:** nginx_ingress_controller_nginx_process_requests_total
+
+**Description:** Total number of client requests
+
+**Type:** counter
+
+**Labels:**
+
+- controller_pod
+
+----------
+
+**Name:** nginx_ingress_controller_nginx_process_resident_memory_bytes
+
+**Description:** Number of bytes of memory in use
+
+**Type:** gauge
+
+**Labels:**
+
+- controller_pod
+
+----------
+
+**Name:** nginx_ingress_controller_nginx_process_virtual_memory_bytes
+
+**Description:** Number of bytes of memory in use
+
+**Type:** gauge
+
+**Labels:**
+
+- controller_pod
+
+----------
+
+**Name:** nginx_ingress_controller_nginx_process_write_bytes_total
+
+**Description:** Number of bytes written
+
+**Type:** counter
+
+**Labels:**
+
+- controller_pod
+
+----------
+
+**Name:** nginx_ingress_controller_request_duration_seconds
+
+**Description:** The request processing time in milliseconds
+
+**Type:** histogram
+
+**Labels:**
+
+- controller_pod
+- ingress_uid
+- method
+- status
+
+----------
+
+**Name:** nginx_ingress_controller_request_size
+
+**Description:** The request length (including request line, header, and request body)
+
+**Type:** histogram
+
+**Labels:**
+
+- controller_pod
+- ingress_uid
+- method
+- status
+
+----------
+
+**Name:** nginx_ingress_controller_requests
+
+**Description:** The total number of client requests.
+
+**Type:** counter
+
+**Labels:**
+
+- controller_pod
+- ingress_uid
+- status
```

Comment:
An additional issue #3818 can be solved by incorporating a status_class:5xx in the metric implemented as part of this proposal